python/ray/experimental/data/dataset.py

```diff
@@ -219,7 +221,92 @@ def split(self, n: int,
         Returns:
             A list of ``n`` disjoint dataset splits.
         """
-        raise NotImplementedError  # P1
+        assert n > 0 and n <= len(self._blocks)
+        assert locality_hints is None or len(locality_hints) == n
+
+        # Split self._blocks into n chunks. the first {num_large_chunks} chunks
+        # have {small_chunk_size + 1} blocks, and remaining
+        # chunks have {small_chunk_size} blocks.
+        small_chunk_size = len(self._blocks) // n
+        num_large_chunks = len(self._blocks) - small_chunk_size * n
+
+        datasets = []
+        block_refs = list(self._blocks)
+
+        if locality_hints is None:
+            cur = 0
+            for i in range(n):
+                end = cur + small_chunk_size
+                if i < num_large_chunks:
+                    end = end + 1
+                datasets.append(Dataset(block_refs[cur:end]))
+                cur = end
+            return datasets
+
+        # If the locality_hints is set, we use a greedy algorithm
+        # to co-locate the blocks with the actors based on block
+        # and actor's node_id:
+        #
+        # 1. We get the node_id where the block resides in and store it in
+        # a map {node_id: [blocks]}. Note that each block could
+        # be located on multiple nodes. For simplicity, we use the first
+        # node_id so that there is one to one mapping from node_id to block
+        # in the map.
+        #
+        # 2. We do two round of allocations. In the first round, we allocate
+        # blocks for each actor based by sustract the blocks from the map
+        # we build from previous step.
+        #
+        # 3. Some actors might have more blocks than needed while the
+        # others don't have enough blocks. We do second round of allocation
+        # for the reminder blocks so that every actor have sufficient blocks.
+
+        # wait might affect the throughput due to stragglers?
+        ray.wait(block_refs)
```

Comment:
Note: Wait has `num_returns=1` by default, so it is returned after "one" ref is ready to go. So you should use ray.get here (or ray.wait(block_refs, num_returns=len(block_refs), fetch_local=False))